### ISSUE TITLE
zsh support

### ISSUE THREAD

**Issue by kubukoz** [cachix/devenv#36]
Especially when used with direnv - it'd be nice to keep a similar UX to what plain Nix with flakes does.

devenv:

<img width="392" alt="image" src="https://user-images.githubusercontent.com/894884/201531361-50a0316a-f959-4170-a3f7-976383b89557.png">

direnv with flakes:

<img width="484" alt="image" src="https://user-images.githubusercontent.com/894884/201531376-d6a707df-4ac8-4884-9696-d044bc7251b6.png">

**Comment by domenkozar** [comment]
Here's where we set `$PS1`: https://github.com/cachix/devenv/blob/main/src/modules/top-level.nix#L81

How should it be adapted for zsh?

**Comment by kubukoz** [comment]
hmm, I think I got confused here. I thought since the prompt says `bash`, we're actually in bash - but this would suggest otherwise:

<img width="575" alt="image" src="https://user-images.githubusercontent.com/894884/201536558-8f225e9d-0c79-4406-80c6-7fefd202ec57.png">

**Comment by shyim** [comment]
I guess with devenv your actual shell is bash and not your configured one. I guess the `SHELL` env gets not overwritten and is from the parent shell

**Comment by edio** [comment]
@domenkozar , I think the author wanted devenv to launch zsh rather than bash. It's not about `PS1`.

We can't really compare to direnv, it's a different beast, a more fair would be a comparison to nix-shell. It too launches `bash` no matter what.

There's a 7-year old issue about supporting shells other than bash in nix-shell. It's still open :see_no_evil: 
https://github.com/NixOS/nix/pull/545

@kubukoz , one way to check your shell from within is
```
echo $0
```

**Comment by manveru** [comment]
Yeah, this is the main thing keeping me from using direnv+devenv over direnv+mkShell... forcing people into using bash isn't an option for me.

**Comment by domenkozar** [comment]
@manveru I'd really like to fix this but I'm not sure how to reproduce the issue.

**Comment by manveru** [comment]
I have this in `.envrc`:

```
watch_file devenv.nix
watch_file devenv.yaml
watch_file devenv.lock
if [[ ! "$DIRENV_ACTIVE" =~ (^|:)"$PWD"(:|$) ]]; then
  devenv shell
fi
```

and it starts a new bash instead of exporting the environment to zsh.

**Comment by domenkozar** [comment]
And this otherwise works with `nix develop`? If I understand correctly, nix always launches bash.

**Comment by manveru** [comment]
Yeah, I don't use `nix develop` when `direnv`+`use flake` is possible if i want to run more than one or two commands. I'd rather not spend time customizing bash to become approximately as usable as my zsh setup.
TBH I'm not sure there's a way to achieve this, and for the time being I'm adopting a combination of `flake-parts`+`mkShell`+`direnv`+`use flake`, which works nicely but obviously requires quite a bit more boilerplate and lacks the immediate support for all those language environments out of the box.

**Comment by domenkozar** [comment]
https://ianthehenry.com/posts/how-to-learn-nix/nix-zshell/ shows just how painful it is to support `zsh`. I'm going to work on #240 to fix this.

**Comment by jasonprado** [comment]
[any-nix-shell](https://github.com/haslersn/any-nix-shell) manages to make `nix-shell -p cowsay` launch a zsh shell for me with only ~2 lines of config in my home-manager setup. It'd be great if launching `devenv shell` did the same! I opened an issue there to discuss: https://github.com/haslersn/any-nix-shell/issues/33

**Comment by rolfschr** [comment]
The following is sufficient for me:

```nix
  enterShell = ''
    exec fish
  '';
```

I also have this for my prompt (fish) specifically for `nix-shell` but it worked also for `devenv shell`:

```fish
  if test -n "$IN_NIX_SHELL"
    set_color cyan; printf "[nix-shell] "; set_color normal
  end
```

**Comment by agj** [comment]
Hello! Ran into this problem today. When using plain Nix flakes, I use `nix develop -c $SHELL` to start it in my own choice of shell. It's easy to put that command in a task runner and make it easy for the whole team. But there's no similar option with `devenv shell` that I could find. Any chance that this flag could be added?

**Comment by Shyrogan** [comment]
> The following is sufficient for me:
> 
>   enterShell = ''
>     exec fish
>   '';
> 
> I also have this for my prompt (fish) specifically for `nix-shell` but it worked also for `devenv shell`:
> 
>   if test -n "$IN_NIX_SHELL"
>     set_color cyan; printf "[nix-shell] "; set_color normal
>   end

While this works, it kinda gets annoying when working with services, running `devenv up` will open a new fish shell instead of opening the terminal UI for services.

**Comment by domenkozar** [comment]
> Hello! Ran into this problem today. When using plain Nix flakes, I use `nix develop -c $SHELL` to start it in my own choice of shell. It's easy to put that command in a task runner and make it easy for the whole team. But there's no similar option with `devenv shell` that I could find. Any chance that this flag could be added?

`devenv shell zsh` should work

**Comment by rseichter** [comment]
@domenkozar  Could it be that easy? I'll have a go at trying `devenv shell zsh`, thanks.

**Comment by dluksza** [comment]
maybe we could have a system-wide configuration in `~/.config/devenv/devenv.yaml` where the default shell could be specified?

**Comment by rseichter** [comment]
> maybe we could have a system-wide configuration in `~/.config/devenv/devenv.yaml` where the default shell could be specified?

Would examining the `$SHELL` environment variable be an option? That might avoid the necessity for a configuration option.

**Comment by domenkozar** [comment]
@sandydoo now that we have custom logic to launch bash, we could also support zsh?

**Comment by mcdonc** [comment]
I created a devenv extension that drops you into a ZSH shell instead of a bash shell here:

https://github.com/mcdonc/devenv-zsh

**Comment by domenkozar** [comment]
> Instead of creating a bunch of options that will need to be overridden by each team member to get their preferred shell working, I think we should probably sniff $SHELL or sniff using getent for the type of shell we should use based on the user's default shell in the passwd file and do the needful to execute that shell within/after enterShell, maybe offering an option to turn that behavior off, and various options to execute shell-specific commands upon first invoking each shell.

This does sacrifice reproducibility, maybe we would be better off setting bash as default on linux and zsh on macos, with the option to override that?

**Comment by mcdonc** [comment]
> This does sacrifice reproducibility, maybe we would be better off setting bash as default on linux and zsh on macos, with the option to override that?

I woulld be fine with that, although once interactive mode starts, I don't know there's much value to repeatability. Even people on the same team may be using a different shell day-to-day, and if the override was a manual setting per-user, they'd each need to manage e.g. a devenv.local.nix.  I don't think any of devenv's competitors require this.

But probably best to start by getting something working cleanly first anyway without the sniffing.

It might be enough to offer up a specialization of devenv.rs' ``exec_in_shell`` that looked inside a config option (instead of command line arguments) for a command that happened to be another interactive shell that bash could start.   Maybe in the config:

subshell = "${pkgs.zsh} -i";

Then devenv would call the specialization function if that value was set?  Would probably need to emit something without the "executing in shell" wrapper:

E.g.

```
• Running  interactive subshell /path/to/zsh/bin/zsh -i ...
.....
```

Instead of ...

```
• Executing in shell ...
.....
✔ Executing in shell in 22.3s

```

But I dunno.  Not sure if that still requires specializing processes up/down code.

**Comment by domenkozar** [comment]
With recent work by @sandydoo and @mcdonc I think we're getting closer to native zsh support in devenv. This is especially important for macOS users <3

**Comment by domenkozar** [comment]
Preliminary work in https://github.com/cachix/devenv/commit/88e4aba045938ee00aa192840460c36f7fc993e2

**Comment by mcdonc** [comment]
FWIW, I added devenv 2.0.X support to https://github.com/mcdonc/devenv-zsh .  If anyone uses it, and has feedback, I'd appreciate it.

**Comment by AmentumStudio** [comment]
> FWIW, I added devenv 2.0.X support to https://github.com/mcdonc/devenv-zsh . If anyone uses it, and has feedback, I'd appreciate it.

Thank you very much!
On `devenv 2.0.6+55d2bb4 (x86_64-linux)` `devenv shell` still runs `bash`, strangely enough this `nix run github:cachix/devenv/auto-reload-shell shell` runs `zsh` with your plugin on, and `bash` with it turned off, so not sure what is happening here.

**Comment by domenkozar** [comment]
https://devenv.sh/blog/2026/05/07/devenv-21-nix-with-zsh-fish-and-nushell-via-libghostty/
